File: union-find/dsu.cpp
#ifndef __STD_VECTOR__
#define __STD_VECTOR__
#include <vector>
#endif

/**
 * https://judge.yosupo.jp/submission/113659
 */
class DSU_Iterative_By_Size
{
private:
    std::vector<int> par;
    std::vector<int> size;

public:
    DSU_Iterative_By_Size(int sz) : par(sz), size(sz)
    {
        for (int i = 0; i < sz; i++)
        {
            par[i] = i;
            size[i] = 1;
        }
    }

    int find(int a)
    {
        int x = a;
        int tmp;
        while (this->par[x] != x)
            x = this->par[x];
        while (this->par[a] != x)
        {
            tmp = this->par[a];
            this->par[a] = x;
            a = tmp;
        }
        return x;
    }

    bool un(int x, int y)
    {
        int tmp;
        int xr = this->find(x), yr = this->find(y);
        if (xr == yr)
            return false;
        else
        {
            if (this->size[xr] < this->size[yr])
            {
                // swap
                tmp = xr;
                xr = yr;
                yr = tmp;
            }
            this->par[yr] = xr;
            this->size[xr] += this->size[yr];
            this->size[yr] = 0;
            return true;
        }
    }
};

File: union-find/dsu.py
class DSU_Recursive_By_Size:
    """Iterative find, rank by size, and path compression"""

    def __init__(self, n):
        self.par = list(range(n))
        self.size = [1] * n

    def find(self, x):
        if self.par[x] != x:
            self.par[x] = self.find(self.par[x])
        return self.par[x]

    def union(self, x, y):
        xr, yr = self.find(x), self.find(y)
        if xr == yr:
            return False
        else:
            if self.size[xr] < self.size[yr]:
                xr, yr = yr, xr
            self.par[yr] = xr
            self.size[xr] += self.size[yr]
            self.size[yr] = 0
            # self.par[self.find(x)] = self.find(y)
            return True


class DSU_Iterative_By_Size:
    """Iterative find, rank by size, and path compression
    https://judge.yosupo.jp/submission/113655"""

    def __init__(self, n):
        self.par = list(range(n))
        self.size = [1] * n

    def find(self, a):
        x = a
        while self.par[x] != x:
            x = self.par[x]
        while self.par[a] != x:
            self.par[a], a = x, self.par[a]
        return x

    def union(self, x, y):
        xr, yr = self.find(x), self.find(y)
        if xr == yr:
            return False
        else:
            if self.size[xr] < self.size[yr]:
                xr, yr = yr, xr
            self.par[yr] = xr
            self.size[xr] += self.size[yr]
            self.size[yr] = 0
            # self.par[self.find(x)] = self.find(y)
            return True
